plugins/libplugin-pay.c

```diff
@@ -356,16 +447,89 @@ static struct command_result *
 payment_waitsendpay_finished(struct command *cmd, const char *buffer,
 			     const jsmntok_t *toks, struct payment *p)
 {
+	struct payment *root;
+	struct channel_hint hint;
+	struct route_hop *hop;
+	assert(p->route != NULL);
+
 	p->result = tal_sendpay_result_from_json(p, buffer, toks);
 
-	if (p->result->state == PAYMENT_COMPLETE)
+	if (p->result->state == PAYMENT_COMPLETE) {
 		p->step = PAYMENT_STEP_SUCCESS;
-	else
-		p->step = PAYMENT_STEP_FAILED;
+		goto cont;
+	}
 
-	/* TODO examine the failure and eventually stash exclusions that we
-	 * learned in the payment, so sub-payments can avoid them. */
+	p->step = PAYMENT_STEP_FAILED;
+	root = payment_root(p);
+
+	switch (p->result->failcode) {
+	case WIRE_PERMANENT_CHANNEL_FAILURE:
+	case WIRE_CHANNEL_DISABLED:
+	case WIRE_UNKNOWN_NEXT_PEER:
+	case WIRE_REQUIRED_CHANNEL_FEATURE_MISSING:
+		/* All of these result in the channel being marked as disabled. */
+		assert(*p->result->erring_index < tal_count(p->route));
+		hop = &p->route[*p->result->erring_index];
+		hint.enabled = false;
+		hint.scid.scid = hop->channel_id;
+		hint.scid.dir = hop->direction;
+		hint.estimated_capacity = AMOUNT_MSAT(0);
+		tal_arr_expand(&root->channel_hints, hint);
```

Comment:
is it possible this channel hint is already in the `channel_hints` list?

Reply:
Indeed it is, if we have two successive attempts with smaller amounts (requires MPP) we could end up adding a new one. We'd exclude due to the smaller one, but adding it twice is suboptimal. Will change this into an `update_or_create`.